# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [236]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import datetime
import seaborn as sns

path="/Users/atousachaharsoughshirazi/desktop/Bootcamp/Unit_6/Step_2/data"
os.chdir(path) 
os.getcwd()

'/Users/atousachaharsoughshirazi/Desktop/Bootcamp/Unit_6/Step_2/data'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [237]:
df=pd.read_csv('dfno.csv')
df.drop(df.columns[0], axis=1, inplace=True)
df.head()

Name       state  summit_elev  vertical_drop  base_elev  \
0     Eaglecrest Ski Area      Alaska         2600           1540       1200   
1        Hilltop Ski Area      Alaska         2090            294       1796   
2     Sunrise Park Resort     Arizona        11100           1800       9200   
3  Boreal Mountain Resort  California         7700            500       7200   
4        Donner Ski Ranch  California         8012            750       7031   

   trams  fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0      0        0.0          0          0     0  ...              640.0   
1      0        0.0          0          0     0  ...               30.0   
2      0        0.0          0          1     2  ...              800.0   
3      0        0.0          0          1     1  ...              380.0   
4      0        0.0          0          0     0  ...              505.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0            60.0              45.0       44.0            350.0          47.0   
1            30.0             150.0       36.0             69.0          30.0   
2            80.0             115.0       49.0            250.0          74.0   
3           200.0             150.0       54.0            400.0          49.0   
4            60.0             163.0       82.0            400.0          75.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0      53.00000               90.0             0.0         0  
1      34.00000              152.0            30.0         0  
2      78.00000              104.0            80.0         1  
3      64.16681              150.0           200.0         2  
4      75.00000              170.0             0.0         2  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [238]:
 # select object type columns
dfo=df['state'].values
df = pd.concat([df.drop(columns='state', axis=1), pd.get_dummies(dfo)], axis=1)
#df = pd.concat([df.drop([df.state], axis=1), pd.get_dummies(df.state)], axis=1)
df.head(-5)

Name  summit_elev  vertical_drop  base_elev  trams  \
0      Eaglecrest Ski Area         2600           1540       1200      0   
1         Hilltop Ski Area         2090            294       1796      0   
2      Sunrise Park Resort        11100           1800       9200      0   
3   Boreal Mountain Resort         7700            500       7200      0   
4         Donner Ski Ranch         8012            750       7031      0   
..                     ...          ...            ...        ...    ...   
95            Grand Geneva         1086            211        875      0   
96         Mount La Crosse         1110            516        594      0   
97         Nordic Mountain         1137            265        872      0   
98                Sunburst         1100            214        866      0   
99             Trollhaugen         1200            260        920      0   

    fastEight  fastSixes  fastQuads  quad  triple  ...  Ohio  Pennsylvania  \
0         0.0          0          0     0       0  ...     0             0   
1         0.0          0          0     0       1  ...     0             0   
2         0.0          0          1     2       3  ...     0             0   
3         0.0          0          1     1       3  ...     0             0   
4         0.0          0          0     0       1  ...     0             0   
..        ...        ...        ...   ...     ...  ...   ...           ...   
95        0.0          0          0     0       0  ...     0             0   
96        0.0          0          0     0       0  ...     0             0   
97        0.0          0          0     0       1  ...     0             0   
98        0.0          0          0     0       0  ...     0             0   
99        0.0          0          0     2       0  ...     0             0   

    Rhode Island  South Dakota  Utah  Vermont  Washington  West Virginia  \
0              0             0     0        0           0              0   
1              0             0     0        0           0              0   
2              0             0     0        0           0              0   
3              0             0     0        0           0              0   
4              0             0     0        0           0              0   
..           ...           ...   ...      ...         ...            ...   
95             0             0     0        0           0              0   
96             0             0     0        0           0              0   
97             0             0     0        0           0              0   
98             0             0     0        0           0              0   
99             0             0     0        0           0              0   

    Wisconsin  Wyoming  
0           0        0  
1           0        0  
2           0        0  
3           0        0  
4           0        0  
..        ...      ...  
95          1        0  
96          1        0  
97          1        0  
98          1        0  
99          1        0  

[100 rows x 56 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [239]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [240]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
X_train

array([[-0.04259049,  1.53574936, -0.36006005, ..., -0.1393466 ,
        -0.32444284, -0.17149859],
       [-0.35078064,  0.64658975, -0.52404842, ..., -0.1393466 ,
        -0.32444284, -0.17149859],
       [-0.52920652, -0.58835416, -0.46889538, ..., -0.1393466 ,
        -0.32444284, -0.17149859],
       ...,
       [-0.58597839, -0.93943107, -0.4600709 , ..., -0.1393466 ,
        -0.32444284, -0.17149859],
       [-0.83090846, -0.50014388, -0.82922858, ..., -0.1393466 ,
        -0.32444284, -0.17149859],
       [-0.67519133, -1.0735107 , -0.5332406 , ..., -0.1393466 ,
        -0.32444284, -0.17149859]])

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [241]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [242]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
y_pred

array([63.37531172, 52.93425748, 53.36945213, 76.39270586, 53.78044998,
       52.90167992, 46.2966565 , 67.01581894, 43.00086336, 63.31603294,
       56.05391872, 67.31605657, 41.05869318, 42.13028905, 59.9104311 ,
       45.35464736, 40.76926571, 39.03067092, 44.68606641, 64.00135523,
       58.20011152, 71.62497907, 67.47273974, 86.97311411, 71.9841871 ,
       78.96178654, 70.9480702 ])

## Review Model Outcomes — Iterate over additional models as needed

In [243]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from . 
# You can plug y_test and y_pred into the functions to evaluate the model

In [244]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)


0.41112396393318684

In [245]:
mean_absolute_error(y_test, y_pred)

8.296712900981388

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [246]:
model.intercept_

58.221272966934194

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [247]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        1.876704e+02
vertical_drop      3.188073e+01
base_elev          1.628977e+02
trams              9.947598e-14
fastEight          1.243450e-13
fastSixes          4.263256e-14
fastQuads          1.278707e+00
quad               2.485373e-01
triple             1.648444e+00
double             1.256650e+00
surface            1.316271e+00
total_chairs       1.065566e+00
Runs               5.078588e+00
TerrainParks       1.117845e+00
LongestRun_mi      1.275669e+00
SkiableTerrain_ac  8.923545e-01
Snow Making_ac     1.682376e+00
daysOpenLastYear   3.679442e+00
yearsOpen          4.160712e-01
averageSnowfall    5.026532e-02
AdultWeekday       9.017691e+00
projectedDaysOpen  3.541612e+00
NightSkiing_ac     5.608360e-01
clusters           1.125691e+01
Alaska             2.088103e+00
Arizona            1.670270e+00
California         7.648262e+00
Colorado           1.784176e+00
Connecticut        1.215724e+00
Idaho              2.618208e+00
Illinois           1.494714e+00
Indiana            2.029597e+00
Iowa               2.500898e+00
Maine              1.542087e-01
Maryland           4.558863e-01
Massachusetts      6.387035e-02
Michigan           3.757049e+00
Minnesota          2.699601e+00
Missouri           8.874743e-01
Nevada             1.333811e+01
New Hampshire      1.337436e+00
New Mexico         1.541610e+00
New York           1.061355e+00
North Carolina     1.366941e+00
Ohio               2.827410e+00
Pennsylvania       8.317122e-01
Rhode Island       1.167873e-27
South Dakota       1.062446e-27
Utah               4.021059e+00
Vermont            1.834896e+00
Washington         1.062446e-27
West Virginia      9.382527e-01
Wisconsin          2.741371e+00
Wyoming            2.565017e+00

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [248]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df

X2=X.iloc[:,:-31]
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 

In [249]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y_train, y_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)
X2_train

array([[-0.04259049,  1.53574936, -0.36006005, ...,  1.09935601,
         0.9230118 , -0.03543655],
       [-0.35078064,  0.64658975, -0.52404842, ...,  0.8471124 ,
         0.79720102,  2.06764551],
       [-0.52920652, -0.58835416, -0.46889538, ...,  0.40044927,
         0.67139024,  0.57122174],
       ...,
       [-0.58597839, -0.93943107, -0.4600709 , ..., -0.66634929,
        -1.67707769,  0.57122174],
       [-0.83090846, -0.50014388, -0.82922858, ..., -0.41410568,
        -0.41896987, -0.23765598],
       [-0.67519133, -1.0735107 , -0.5332406 , ..., -1.23389743,
        -1.25770842,  0.16678288]])

In [250]:
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [251]:
y_pred=model.predict(X2_test)
y_pred

array([61.12925998, 61.260308  , 58.74406795, 73.55959699, 53.60307313,
       56.03755352, 43.80043147, 81.52666927, 52.83191099, 62.37028746,
       59.09140031, 56.84865941, 39.65297553, 51.54418395, 65.41578391,
       70.22588662, 63.48478845, 57.60368566, 50.57100921, 58.9385423 ,
       69.61047789, 78.8663824 , 63.88875524, 91.10461842, 80.46099694,
       84.47177317, 74.93060903])

In [252]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)

0.7458560699280465

In [253]:
mean_absolute_error(y_test, y_pred)

5.6678825013084975

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [254]:
model.intercept_

59.5804546416141

In [255]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])



Coefficient
summit_elev        6.444221e+00
vertical_drop      8.093708e-01
base_elev          6.829685e+00
trams              6.439294e-15
fastEight          2.664535e-15
fastSixes          8.881784e-15
fastQuads          6.034601e-01
quad               5.080437e-01
triple             1.534494e+00
double             8.341738e-01
surface            1.191113e+00
total_chairs       8.815044e-01
Runs               1.597010e+00
TerrainParks       9.098236e-01
LongestRun_mi      2.204053e+00
SkiableTerrain_ac  3.279765e+00
Snow Making_ac     6.804519e-02
daysOpenLastYear   3.259594e+00
yearsOpen          2.292385e-01
averageSnowfall    3.401974e-01
AdultWeekday       9.306101e+00
projectedDaysOpen  2.088829e+00
NightSkiing_ac     8.329976e-02

#### Model 3

In [256]:
X2.drop(['summit_elev','base_elev'],axis=1, inplace=True)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 

y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler.transform(X2) 


In [257]:
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y_train, y_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)
X2_train

array([[ 1.53574936,  0.        ,  0.        , ...,  1.09935601,
         0.9230118 , -0.03543655],
       [ 0.64658975,  0.        ,  0.        , ...,  0.8471124 ,
         0.79720102,  2.06764551],
       [-0.58835416,  0.        ,  0.        , ...,  0.40044927,
         0.67139024,  0.57122174],
       ...,
       [-0.93943107,  0.        ,  0.        , ..., -0.66634929,
        -1.67707769,  0.57122174],
       [-0.50014388,  0.        ,  0.        , ..., -0.41410568,
        -0.41896987, -0.23765598],
       [-1.0735107 ,  0.        ,  0.        , ..., -1.23389743,
        -1.25770842,  0.16678288]])

In [258]:
lm = linear_model.LinearRegression()
model = lm.fit(X2_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [259]:
y_pred=model.predict(X2_test)
y_pred

array([61.76743883, 61.69720073, 58.57029192, 72.96584123, 53.05409077,
       55.23213641, 43.58334766, 80.79946186, 52.07987274, 62.49395478,
       59.31279759, 56.60890412, 40.04254563, 52.87284366, 64.48729603,
       67.67464505, 64.49299406, 59.0511567 , 49.75320232, 59.26579085,
       68.35373703, 79.03668386, 63.54052007, 90.9610448 , 83.72634504,
       86.90813595, 75.15003667])

In [260]:
explained_variance_score(y_test, y_pred)

0.7394399198373897

In [261]:
mean_absolute_error(y_test, y_pred)

5.800952059308963

In [262]:
model.intercept_

59.59863206236464

In [263]:
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])

Coefficient
vertical_drop      1.956593e+00
trams              4.440892e-16
fastEight          1.332268e-15
fastSixes          5.551115e-16
fastQuads          4.013829e-01
quad               5.087160e-01
triple             1.323268e+00
double             8.850826e-01
surface            1.066666e+00
total_chairs       8.249113e-01
Runs               1.424794e+00
TerrainParks       1.019371e+00
LongestRun_mi      2.404189e+00
SkiableTerrain_ac  3.599091e+00
Snow Making_ac     2.384378e-01
daysOpenLastYear   3.191578e+00
yearsOpen          2.812538e-01
averageSnowfall    9.608801e-02
AdultWeekday       9.381038e+00
projectedDaysOpen  2.163052e+00
NightSkiing_ac     2.249031e-01

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.41112396393318684 | 8.296712900981388 |-|
| Model 2. | 0.7458560699280465| 5.6678825013084975 |'state'|
| Model 3. | 0.7394399198373897 | 5.800952059308963|'state','summit_elev','base_elev'|


Model Selection:

In [264]:
#Model 2